client/modules/components/src/Wikitext.svelte

```diff
@@ -0,0 +1,147 @@
+<!--
+  @component Generic wikitext container.
+-->
+<script lang="ts">
+  import * as FTML from "ftml-wasm-worker"
+  import { createAnimQueued, createMutatingLock, perfy, toFragment } from "wj-util"
+  import Card from "./Card.svelte"
+  import morphdom from "morphdom"
+  import Spinny from "./Spinny.svelte"
+  import { anim } from "./lib/animation"
+
+  type Rendered = { html: string; styles: string[] }
+  type WikitextInput =
+    | Promisable<string>
+    | Promisable<Rendered>
+    | (() => Promisable<string>)
+    | (() => Promisable<Rendered>)
+
+  /**
+   * Wikitext input. May be given in quite a few ways:
+   * 1. As a string.
+   * 2. As an object, `{ html: string, style: string }`.
+   * 3. As a function returning either a string or the aformentioned object.
+   * 4. As a `Promise` resolving to either a string or the aforementioned object.
+   * 5. As a function returning a `Promise` like the aformentioned one.
+   *
+   * Providing the input as a function has performance benefits. If the wikitext
+   * takes a long time to render, and the source wikitext is frequently updating,
+   * providing a function will make it so that the wikitext is only evalulated when
+   * it needs to be. Otherwise, the wikitext would be evalulated constantly, and that
+   * has obvious performance implications.
+   */
+  export let wikitext: WikitextInput = ""
+
+  /**
+   * Flags the component to morph the output container
+   * rather than entirely replacing the contents.
+   *
+   * This is for performance reasons. If you are frequently updating the wikitext,
+   * it's probably best if the DOM is morphed rather than replaced.
+   */
+  export let morph = false
+
+  let element: HTMLElement
+  let stylesheet: HTMLStyleElement
+
+  let rendering = false
+
+  let perfRender = 0
+
+  const render = createMutatingLock(async (wikitext: WikitextInput) => {
+    const displayIndicatorTimeout = setTimeout(() => (rendering = true), 100)
+    const measure = perfy()
+    if (typeof wikitext === "function") wikitext = wikitext()
+    wikitext = await wikitext
```

Comment:
Yeah, it's a bit JS wizardary heavy. Basically, it's very quickly "decomposing" the wikitext variable into its final form, as it may or may not be a Promise, and it may or may not be a function, and that function may return a promise, etc. One thing to note about JS `await` is that you can await anything and it's fine. If it's not a promise or thenable it just resolves to the value.

Reply:
I was more wondering why it would be in different forms like that to begin with.

Reply:
Makes it very easy to use. If you have wikitext already, you can provide it. If you have a Promise for wikitext, you can just provide it directly without shoving any async into markup or into your scripts. If it's a function, you can use that for performance reasons.

Because the component is also supposed to be highly reusable and versatile, it makes sense to have a very flexible input